Add bytes() to Blob and PushMessageData

annevk opened the pull request:
#### d35627646eb3b30079f2ec5bbbc75b02b2eb4e2b
<pre>
Add bytes() to Blob and PushMessageData
<a href="https://bugs.webkit.org/show_bug.cgi?id=274119">https://bugs.webkit.org/show_bug.cgi?id=274119</a>
<a href="https://rdar.apple.com/128418858">rdar://128418858</a>

Reviewed by Youenn Fablet.

This implements <a href="https://github.com/w3c/FileAPI/pull/198">https://github.com/w3c/FileAPI/pull/198</a> and
<a href="https://github.com/w3c/push-api/pull/370">https://github.com/w3c/push-api/pull/370</a> creating parity for these
APIs with Fetch&apos;s Body.

This incorporates the tests from
<a href="https://github.com/web-platform-tests/wpt/pull/46232">https://github.com/web-platform-tests/wpt/pull/46232</a> modulo a typo fix.
PushMessageData test coverage is done through a local test that would
be good to upstream at some point.

* LayoutTests/http/wpt/push-api/pushEvent.any.js:
(test):
* LayoutTests/imported/w3c/web-platform-tests/FileAPI/Blob-methods-from-detached-frame-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/FileAPI/Blob-methods-from-detached-frame.html:
* LayoutTests/imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.js: Added.
(string_appeared_here.promise_test.async const):
* LayoutTests/imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/FileAPI/blob/w3c-import.log:
* Source/WebCore/Modules/push-api/PushMessageData.cpp:
(WebCore::PushMessageData::arrayBuffer):
(WebCore::PushMessageData::bytes):
* Source/WebCore/Modules/push-api/PushMessageData.h:
* Source/WebCore/Modules/push-api/PushMessageData.idl:
* Source/WebCore/fileapi/Blob.cpp:
(WebCore::Blob::arrayBuffer):
(WebCore::Blob::bytes):
* Source/WebCore/fileapi/Blob.h:
* Source/WebCore/fileapi/Blob.idl:

Canonical link: <a href="https://commits.webkit.org/279263@main">https://commits.webkit.org/279263@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/841b7d4dd8c1fcacf53234c6ce7759865cf009d1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52962 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32299 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5449 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56241 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3685 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55267 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/39107 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3411 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42964 "Passed tests") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/2373 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55060 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/30037 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45719 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24088 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/27087 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3038 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1844 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/48943 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3195 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57836 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28103 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3159 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/50362 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29323 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45935 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/49662 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11560 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30242 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29077 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/fileapi/Blob.cpp:
```diff
@@ -304,6 +304,27 @@ void Blob::arrayBuffer(Ref<DeferredPromise>&& promise)
+        }
+        auto view = Uint8Array::tryCreate(WTFMove(arrayBuffer), 0, arrayBuffer->byteLength());
+        if (!view) {
+            promise->reject(Exception { ExceptionCode::InvalidStateError });
```
InvalidStateError or OutOfMemoryError?
We should look at what we are doing for arrayBuffer.

Ditto for `if (!arrayBuffer)`

annevk replied:
This is literally copied from the code for `::arrayBuffer()` above. Could maybe attempt to abstract it so it's shared.

marcoscaceres commented on LayoutTests/imported/w3c/web-platform-tests/FileAPI/Blob-methods-from-detached-frame.html:
```diff
@@ -48,6 +49,14 @@
+
+    const charCodeBytes = await bytes.call(blob);
+    assert_equals(charCodeArrayToString(charCodeBytes), "bar");
+}, "bytes()");
```
Nit (and probably a followup)... it might be nicer to have more complete descriptions of exactly what the aim of the test is, though I can see the rest of the tests are also just the method names.

marcoscaceres commented on LayoutTests/imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.js:
```diff
@@ -0,0 +1,45 @@
+  const uint8array = await blob.bytes();
+  assert_true(uint8array instanceof Uint8Array);
+  assert_equals_typed_array(uint8array, input_arr);
+}, "Blob.bytes()")
```
Maybe, just to complement the non-ascii one below... also, more correct to say `.prototype.` so it's clear from the description it's not testing a class-side static: 

```suggestion
}, "Blob.prototype.bytes() with ascii input")
```

marcoscaceres commented on LayoutTests/imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.js:
```diff
@@ -0,0 +1,45 @@
+promise_test(async () => {
+  const input_arr = new TextEncoder().encode("PASS");
+  const blob = new Blob([input_arr]);
+  const uint8array_results = await Promise.all([blob.bytes(),
+      blob.bytes(), blob.bytes()]);
```
Nit on whitespace... maybe?: 

```suggestion
  const uint8array_results = await Promise.all(
    [blob.bytes(), blob.bytes(), blob.bytes()]
  );
```

marcoscaceres reviewed (approved):
Non-reviewer review of the tests... couple of small suggestions/nits, otherwise looks good.

youennf commented on Source/WebCore/fileapi/Blob.cpp:
```diff
@@ -288,19 +289,38 @@ void Blob::text(Ref<DeferredPromise>&& promise)
+    RefPtr arrayBuffer = blobLoader.arrayBufferResult();
+    if (!arrayBuffer)
+        return Exception { ExceptionCode::InvalidStateError };
+    return *arrayBuffer;
```
Return a reference is a bit dangerous.
In theory, it could return a dead reference if blobLoader or child was not keeping a ref to the array buffer.
Let's return ExceptionOr<Ref<JSC::ArrayBuffer>> or ExceptionOr<RefPtr<>> if the binding generator prefers

youennf commented on Source/WebCore/fileapi/Blob.cpp:
```diff
@@ -288,19 +289,38 @@ void Blob::text(Ref<DeferredPromise>&& promise)
-        auto arrayBuffer = blobLoader.arrayBufferResult();
-        if (!arrayBuffer) {
-            promise->reject(Exception { ExceptionCode::InvalidStateError });
+        promise->resolve<IDLArrayBuffer>(arrayBuffer.releaseReturnValue());
```
We can create a DOMPromiseDeferred<IDLArrayBuffer> from Ref<DeferredPromise>&& which could allow us to use `settle`. We may be able to do something like:
```
void Blob::arrayBuffer(DOMPromiseDeferred<IDLArrayBuffer>&& promise)
{
  ...
      promise.settle(arrayBufferFromBlobLoader(blobLoader));
  ...
}
```

youennf commented on Source/WebCore/fileapi/Blob.cpp:
```diff
@@ -288,19 +289,38 @@ void Blob::text(Ref<DeferredPromise>&& promise)
         }
-        promise->resolve<IDLArrayBuffer>(*arrayBuffer);
+        Ref view = Uint8Array::create(arrayBuffer.releaseReturnValue());
+        promise->resolve<IDLUint8Array>(WTFMove(view));
```
Ditto here, or maybe we could try to add settle to DeferredPromise directly

youennf commented on Source/WebCore/fileapi/Blob.cpp:
```diff
@@ -288,19 +289,33 @@ void Blob::text(Ref<DeferredPromise>&& promise)
+    RefPtr arrayBuffer = blobLoader.arrayBufferResult();
+    if (!arrayBuffer)
+        return Exception { ExceptionCode::InvalidStateError };
+    return Ref { *arrayBuffer };
```
```suggestion
    return arrayBuffer.releaseNonNull();
```

annevk committed:
Add bytes() to Blob and PushMessageData
https://bugs.webkit.org/show_bug.cgi?id=274119
rdar://128418858

Reviewed by Youenn Fablet.

This implements https://github.com/w3c/FileAPI/pull/198 and
https://github.com/w3c/push-api/pull/370 creating parity for these
APIs with Fetch's Body.

This incorporates the tests from
https://github.com/web-platform-tests/wpt/pull/46232 modulo a typo fix.
PushMessageData test coverage is done through a local test that would
be good to upstream at some point.

* LayoutTests/http/wpt/push-api/pushEvent.any.js:
(test):
* LayoutTests/imported/w3c/web-platform-tests/FileAPI/Blob-methods-from-detached-frame-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/FileAPI/Blob-methods-from-detached-frame.html:
* LayoutTests/imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.js: Added.
(string_appeared_here.promise_test.async const):
* LayoutTests/imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/FileAPI/blob/w3c-import.log:
* Source/WebCore/Modules/push-api/PushMessageData.cpp:
(WebCore::PushMessageData::arrayBuffer):
(WebCore::PushMessageData::bytes):
* Source/WebCore/Modules/push-api/PushMessageData.h:
* Source/WebCore/Modules/push-api/PushMessageData.idl:
* Source/WebCore/fileapi/Blob.cpp:
(WebCore::Blob::arrayBuffer):
(WebCore::Blob::bytes):
* Source/WebCore/fileapi/Blob.h:
* Source/WebCore/fileapi/Blob.idl:

Canonical link: https://commits.webkit.org/279263@main
